RHIDP-4807: Third-party plugin installation in RHDH (#733)
redhat-developer/red-hat-developers-documentation-rhdh: modules/dynamic-plugins/ref-example-third-party-plugin-installation.adoc

[id="ref-example-third-party-plugin-installation_{context}"]
= Example of installing a third-party plugin in {product}

> This section provides step-by-step instructions for integrating the link:https://github.com/backstage/community-plugins/tree/main/workspaces/todo/plugins[Todo plugin] into your {product-short}.

Reviewer: "step-by-step" denotes a procedure module

Author: This section provides example steps for users to reference, rather than a set of instructions to follow. Therefore, I think this should be considered a reference module.

Author: Changed the language

> . *Obtain the third-party plugin source code*: Clone the plugins repository and navigate to the link:https://github.com/backstage/community-plugins/tree/main/workspaces/todo/plugins[Todo plugin] directory:

Reviewer: + install the Node.js package dependencies

Author: This section doesn't have any prerequisites, as it is a reference module.

+
--
.Obtain the third-party plugin source code

> [source,bash]

Reviewer: ```suggestion
[source,terminal]
```

----
$ git clone https://github.com/backstage/community-plugins
$ cd community-plugins/workspaces/todo
$ yarn install
----
--

. *Export backend and front-end plugins*: Run the following commands to build the backend plugin, adjust package dependencies for dynamic loading, and generate self-contained configuration schema:
+
--

> .Export a backend plugin

Reviewer: ```suggestion
.Export the backend plugin
```

> [source,bash]

Reviewer: ```suggestion
[source,terminal]
```

----
$ cd todo-backend
$ npx @janus-idp/cli@latest package export-dynamic-plugin
----

> .Output of exporting the backend plugin commands

Reviewer 2: ```suggestion
.Output the export backend plugin command
```

> [source,bash]

Reviewer: ```suggestion
[source,terminal]
```

> .Output of exporting a backend plugin commands
> [source,bash]
> ----
> Building main package
>   executing     yarn build ✔
> Packing main package to dist-dynamic/package.json
> Customizing main package in dist-dynamic/package.json for dynamic loading
>   moving @backstage/backend-common to peerDependencies
>   moving @backstage/backend-openapi-utils to peerDependencies
>   moving @backstage/backend-plugin-api to peerDependencies
>   moving @backstage/catalog-client to peerDependencies
>   moving @backstage/catalog-model to peerDependencies
>   moving @backstage/config to peerDependencies
>   moving @backstage/errors to peerDependencies
>   moving @backstage/integration to peerDependencies
>   moving @backstage/plugin-catalog-node to peerDependencies
> Installing private dependencies of the main package
>    executing     yarn install --no-immutable ✔
> Validating private dependencies
> Validating plugin entry points
> Saving self-contained config schema in /Users/user/Code/community-plugins/workspaces/todo/plugins/todo-backend/dist-dynamic/dist/configSchema.json
> ----
> 

Reviewer: Consider omitting the output (fluff + not maintainable).

```suggestion
```

Reviewer 3: The output in the examples can be useful for users, to help them understand what is going on.

Author: @themr0c Since this is a reference module, we are providing examples for users to refer to.

> You can run the following commands to set default dynamic UI configurations, create front-end plugin assets, and to generate a configuration schema for a front-end plugin:
> 
> .Export a front-end plugin

Reviewer: ```suggestion
. Run the following commands to set default dynamic UI configurations, create front-end plugin assets, and to generate a configuration schema for the front-end plugin:
+
.Export the front-end plugin
```

> [source,bash]

Reviewer: ```suggestion
[source,terminal]
```

----
$ cd ../todo
$ npx @janus-idp/cli@latest package export-dynamic-plugin
----

> .Output of exporting the front-end plugin commands

Reviewer 2: ```suggestion
.Output of the export front-end plugin command
```

> [source,bash]

Reviewer: ```suggestion
[source,terminal]
```

----
No scalprum config. Using default dynamic UI configuration:
{
  "name": "backstage-community.plugin-todo",
  "exposedModules": {
    "PluginRoot": "./src/index.ts"
  }
}
If you wish to change the defaults, add "scalprum" configuration to plugin "package.json" file, or use the '--scalprum-config' option to specify an external config.
Packing main package to dist-dynamic/package.json
Customizing main package in dist-dynamic/package.json for dynamic loading
Generating dynamic frontend plugin assets in /Users/user/Code/community-plugins/workspaces/todo/plugins/todo/dist-dynamic/dist-scalprum
  263.46 kB  dist-scalprum/static/1417.d5271413.chunk.js
...
...
...
  250 B      dist-scalprum/static/react-syntax-highlighter_languages_highlight_plaintext.0b7d6592.chunk.js
Saving self-contained config schema in /Users/user/Code/community-plugins/workspaces/todo/plugins/todo/dist-dynamic/dist-scalprum/configSchema.json
----
--

> 

Reviewer: and subsequent ones

. *Package and publish a third-party plugin*: Run the following commands to navigate to the workspace directory and package the dynamic plugin to build the OCI image:
+
--
.Build an OCI image

> [source,bash]

Reviewer: ```suggestion
[source,terminal]
```

----
$ cd ../..
$ npx @janus-idp/cli@latest package package-dynamic-plugins --tag quay.io/user/backstage-community-plugin-todo:v0.1.1
----

> .Output of building an OCI image commands

Reviewer 2: ```suggestion
.Output of the build OCI image command
```

> [source,bash]

Reviewer: ```suggestion
[source,terminal]
```

    ✔
Successfully built image quay.io/user/backstage-community-plugin-todo:v0.1.1 with following plugins:
  backstage-community-plugin-todo
  backstage-community-plugin-todo-backend-dynamic

Here is an example dynamic-plugins.yaml for these plugins:

plugins:
  - package: oci://quay.io/user/backstage-community-plugin-todo:v0.1.1!backstage-community-plugin-todo
    disabled: false
  - package: oci://quay.io/user/backstage-community-plugin-todo:v0.1.1!backstage-community-plugin-todo-backend-dynamic
    disabled: false
----

.Push the OCI image to a container registry:

> [source,bash]

Reviewer: ```suggestion
[source,terminal]
```

----

> $ podman push quay.io/user/backstage-community-plugin-todo:v0.1.1

Reviewer: Consider making the user-replaced content visible (`_<user>_`)

Reviewer 3: This is an example, not instruction step. I think that using real values like this makes sense here.

----

.Output of pushing the OCI image command

> [source,bash]

Reviewer: ```suggestion
[source,terminal]
```

[id="ref-example-third-party-plugin-installation_{context}"]
= Example of installing a third-party plugin in {product}

This section provides step-by-step instructions for integrating the link:https://github.com/backstage/community-plugins/tree/main/workspaces/todo/plugins[Todo plugin] into your {product-short}.

. *Obtain the third-party plugin source code*: Clone the plugins repository and navigate to the link:https://github.com/backstage/community-plugins/tree/main/workspaces/todo/plugins[Todo plugin] directory:
+
--

> .Example commands to obtain the third-party plugin source code

Reviewer 3: Do we need this "Example commands ..." before every code block here? 
This whole document is called Example, it should be clear that the commands here are examples

[source,bash]
----
$ git clone https://github.com/backstage/community-plugins
$ cd community-plugins/workspaces/todo
$ yarn install
----
--

. *Export backend and front-end plugins*: Run the following commands to build the backend plugin, adjust package dependencies for dynamic loading, and generate self-contained configuration schema:
+
--
.Example commands to export a backend plugin
[source,bash]
----
$ cd todo-backend

> $ npx @janus-idp/cli@latest package export-dynamic-plugin

Reviewer 3: In the original document, I've included in the examples the output of the commands.
Is there a reason why it can't be output here?

Author: I’ve added outputs to the module. To help users distinguish between commands and outputs, I’ve organized them into separate code blocks.
